overtime/algorithms/edgeDeletion.py

```diff
@@ -0,0 +1,285 @@
+        # else if edge's duration start is greater than or equal to the end time of the graph's timespan.
+        if edge.end <= end and edge.start >= departure.time:
+            # if edge's time duration end is less than the destination node's foremost time.
+            if edge.end < destination.time:
+                # add this edge to the foremost tree.
+                temp = tree.edges.add(edge.source.label, edge.sink.label, tree.nodes, edge.start, edge.end)
+                # update the destination node's foremost time.
+                destination.time = edge.end
+                # update the destination node's data.
+                destination.data['foremost_time'] = edge.end
+
+                # update accumulator
+                accumulator += 1
+                if accumulator == h:
+                    break
+
+        elif edge.start >= end:
+            # stop the algorithm.
+            break
+
+    # return the resulting foremost tree.
+    return tree
+
+
+def h_approximation(graph, h):
+    """
+        A h-approximation algorithm to return a set of edges E_ such that (G,λ)\E_ has temporal reachability at most h.
+
+        Parameter(s):
+        -------------
+        graph : TemporalDiGraph
+                A directed, temporal graph.
+        h : int
+            the maximum permitted temporal reachability.
+
+        Returns:
+        --------
+        E_ : list
+            a set of edges such that (G,λ)/E_ has temporal reachability at most h.
+
+    """
+
+    E_ = []
+
+    # find the node whose temporal reachability is more than h
+    while (True):
+        root = ''
+        for node in graph.nodes.set:
+            reachability = ot.calculate_reachability(graph, node.label)
+            if reachability > h:
+                root = node.label
+                break
+
+        # check if the specified root actually exists in the graph.
+        if root == '':
+            print('there is no node whose reachability is more than {}'.format(h))
+            break
+
+        # generate a reachable subtree based on the root
+        subtree = reachable_subtree(graph, root, h)
+
+        E_.extend(subtree.edges.uids())
+        # update (G, λ) ← (G, λ) \ E_
+        for edge in subtree.edges.uids():
+            graph.remove_edge(edge)
+
+    return E_
+
+def max_reachability(graph):
+    """
+        A method to calculate the max temporal reachability of a network.
+
+        Parameter(s):
+        -------------
+        graph : TemporalDiGraph
+                A directed, temporal graph.
+
+        Returns:
+        --------
+        maxReachability : int
+            the max temporal reachability of a network.
+
+    """
+
+    maxReachability = 0
+
+    # check the temporal reachability of each node
+    for node in graph.nodes.labels():
+        r = ot.calculate_reachability(graph, node)
+        if r > maxReachability:
+            maxReachability = r
+
+    return maxReachability
+
+
+def max_endtime(graph):

```

Comment:
1) I think the better place for this function is within class TemporalGraph. ()
2) Also for the implementation you could use 
    timespan = graph.edges.timespan()  # graph timespan.
    maxEndtime = timespan[-1]  # end time.

@soca-git  what do you think?

Reply:
I only saw this now!
I agree with both observations.